### PR TITLE
datetime format and database expression on column level

### PR DESCRIPTION
**Pull request by yxjames**
This PR is a revision of #506 .
Now users can edit datetime format and database expression per column. Without specifying anything, Caravel follows default pandas timestamp format and empty database_expression.
screen shot 2016-06-21 at 12 35 17 pm
If timestamp is stored in epoch format in Integer type, users should input this:
screen shot 2016-06-21 at 12 40 20 pm
If it is stored in date/timestamp type, such as `timestamp with time zone` in postgres, users should input this:
screen shot 2016-06-21 at 12 45 04 pm
Query constructed by Caravel looks like this:
screen shot 2016-06-21 at 12 48 02 pm

### CODE REVIEW

File: caravel/views.py [hunk 1]
```diff
@@ -195,6 +195,22 @@ class TableColumnInlineView(CompactCRUDMixin, CaravelModelView):  # noqa
...
+            "<a href='https://docs.python.org/2/library/"
+            "datetime.html#strftime-strptime-behavior'>"
+            "python datetime string pattern</a> "
+            "expression. If time is stored in epoch "
+            "format, put `epoch`. Leave `Database Expression` "
+            "below empty if timestamp is stored in "
+            "String or Integer(epoch) type"), True),
+        'database_expression': utils.markdown(
+            "The database expression to cast internal datetime "
+            "constants to database date/imestamp type according to the DBAPI. "
```
Comment (georgeke) [inline caravel/views.py]: small typo

File: caravel/models.py [hunk 2]
```diff
@@ -601,6 +584,24 @@ def sql_url(self):
     def sql_link(self):
         return '<a href="{}">SQL</a>'.format(self.sql_url)
 
+    def dttm_converter(self, dttm, tf=None, db_expr=None):
```
Comment (mistercrunch) [inline caravel/models.py]: `self` isn't used in this method, but both `tf` and `db_expr` live in the column object. Maybe this method should be a method of the column itself, and receive only `dttm`.

What about renaming the method `dttm_sql_literal`?

File: caravel/models.py [hunk 3]
```diff
@@ -444,24 +444,6 @@ def grains(self):
             if self.sqlalchemy_uri.startswith(db_type):
                 return grains
 
-    def dttm_converter(self, dttm):
-        """Returns a string that the database flavor understands as a date"""
-        default = "'{}'".format(dttm.strftime('%Y-%m-%d %H:%M:%S.%f'))
-        iso = dttm.isoformat()
-        d = {
```
Comment (mistercrunch) [inline caravel/models.py]: These per-database defaults is what made things work magically before. I feel like these must still be taken into account or this PR will create a regression on the databases that don't use the global default

File: caravel/viz.py [hunk 4]
```diff
@@ -147,14 +147,36 @@ def get_df(self, query_obj=None):
...
+        # datetime format. If no python_date_format is specified, the pattern will
+        # be considered as the default ISO date format
+        # If the datetime format is unix, the parse will use the corresponding
+        # parsing logic.
         if df is None or df.empty:
             raise Exception("No data, review your incantations!")
         else:
             if 'timestamp' in df.columns:
-                df.timestamp = pd.to_datetime(df.timestamp, utc=False)
+                if timestamp_format == "epoch":
```
Comment (mistercrunch) [inline caravel/viz.py]: We may also need support for `epoch_ms` (milliseconds) as it is fairly common

File: caravel/viz.py [hunk 5]
```diff
@@ -147,14 +147,36 @@ def get_df(self, query_obj=None):
         self.results = None
 
         # The datasource here can be different backend but the interface is common
```
Comment (mistercrunch) [inline caravel/viz.py]: this comment is at the wrong spot now

File: caravel/models.py [hunk 6]
```diff
@@ -673,18 +674,25 @@ def query(  # sqla
                 select_exprs += [timestamp_grain]
                 groupby_exprs += [timestamp_grain]
 
-            tf = '%Y-%m-%d %H:%M:%S.%f'
+            column_info = db.session().query(TableColumn).filter_by(
```
Comment (mistercrunch) [inline caravel/models.py]: no need to re-query the db here, there's a dict named `cols` in scope here, so `dttm_col = cols.get(granularity)`. You can do that earlier in this method and use `dttm_col` at line 661 as well.

Assuming `col.dttm_sql_literal` will simplify the rest of the changes here.

File: caravel/viz.py [hunk 7]
```diff
@@ -147,14 +147,36 @@ def get_df(self, query_obj=None):
...
+        timestamp_format = None
+        if self.datasource.type == 'table':
+            dttm_cols = self.datasource.dttm_cols
```
Comment (mistercrunch) [inline caravel/viz.py]: Assuming a `datasource.get_col` (that would need to be created), can't this whole block be as simple as 
`timestamp_format = datasource.get_col(query_obj['granularity']).python_date_format `

File: caravel/viz.py [hunk 8]
```diff
@@ -147,14 +147,36 @@ def get_df(self, query_obj=None):
...
             raise Exception("No data, review your incantations!")
         else:
             if 'timestamp' in df.columns:
-                df.timestamp = pd.to_datetime(df.timestamp, utc=False)
+                if timestamp_format == "epoch":
+                    df.timestamp = pd.to_datetime(
+                        df.timestamp, utc=False, unit="s")
+                else:
+                    df.timestamp = pd.to_datetime(
+                        df.timestamp, utc=False, format=timestamp_format)
```
Comment (mistercrunch) [inline caravel/viz.py]: Does this work for databases that do have a DATETIME or DATE type? Seems like it would only work for strings.

**Comment by mistercrunch** [comment]
You may want to squash these commits as well. Having so many commits increases the risk of merge conflicts on your side.

File: caravel/viz.py (hunk 8 above)
Comment (yxjames) [inline caravel/viz.py]: It works for TIMESTAMP WITHOUT TIME ZONE and DATE types in postgres. I think it should also work for other date type in other sql database

File: caravel/models.py (hunk 3 above)
Comment (yxjames) [inline caravel/models.py]: Moved this logic to the new dttm_sql_literal method

File: caravel/viz.py [hunk 9]
```diff
@@ -146,15 +146,34 @@ def get_df(self, query_obj=None):
...
-                df.timestamp = pd.to_datetime(df.timestamp, utc=False)
+                if timestamp_format == "epoch_s":
+                    df.timestamp = pd.to_datetime(
+                        df.timestamp, utc=False, unit="s")
+                elif timestamp_format == "epoch_ms":
+                    df.timestamp = pd.to_datetime(
+                        df.timestamp, utc=False, unit="ms")
+                else:
+                    df.timestamp = pd.to_datetime(
+                        df.timestamp, utc=False, format=timestamp_format)
```
Comment (mistercrunch) [inline caravel/viz.py]: Does this work when against a DATE or DATETIME type?

**Comment by mistercrunch** [comment]
It's probably time to bring in a test dataset in `caravel/data/__init__.py` with some DATE/DATETIME types (create the table using SqlAlchemy), some string with different format and some epochs. It probably needs to have just one or a few rows, and test that it works against the different databases (the build matrix has postgres, mysql and sqlite already, so that should cover most of it).

File: caravel/viz.py (hunk 9 above)
Comment (yxjames) [inline caravel/viz.py]: Yes it does. Against a DATE or DATETIME type, timestamp_format should be None.

File: caravel/data/__init__.py [hunk 10]
```diff
@@ -1020,3 +1020,84 @@ def load_long_lat_data():
...
+        'ds2': [None, None],
+        'epoch_s': ['epoch_s', None],
+        'epoch_ms': ['epoch_ms', None],
+        'string2': ['%Y%m%d-%H%M%S', None],
+        'string1': ['%Y-%m-%d^%H:%M:%S', None],
+        'string0': ['%Y-%m-%d %H:%M:%S.%f', None],
+        'string3': ['%Y/%m/%d%H:%M:%S.%f', None],
+    }
+    for col in obj.table_columns:
+        print(col.column_name)
```
Comment (mistercrunch) [inline caravel/data/__init__.py]: let's remove this line

File: caravel/data/__init__.py [hunk 11]
```diff
@@ -1020,3 +1020,84 @@ def load_long_lat_data():
...
+        col.python_date_format = dttm_and_expr[0]
+        col.dbatabase_expr = dttm_and_expr[1]
+    db.session.merge(obj)
+    db.session.commit()
+    obj.fetch_metadata()
+    tbl = obj
+
+    print("Creating some slices")
+    i = 0
+    for col in tbl.table_columns:
```
Comment (mistercrunch) [inline caravel/data/__init__.py]: nit: `for i, col in enumerate(tbl.table_columns):`

**Comment by mistercrunch** [comment]
Nice. This is looking solid. I'm going to test it against our production DB when I have a moment and report back.

**Comment by mistercrunch** [comment]
Ok. Everything looks good. This was pretty tricky but I think we support most use cases at this point. Thanks for bearing with me.
![img] (https://media.giphy.com/media/26BRqOTqZob11VVde/giphy.gif)

**Comment by yxjames** [comment]
Thanks for you patience for showing me the right way to write python code! I haven't fix those small issues you just pointed out. You will do it yourself?

**Comment by mistercrunch** [comment]
Those were pretty minor. Submit a PR if you feel like it, otherwise we'll get to it at some point.

**Pull request merged**